compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -2232,24 +2233,49 @@ public void visit(BLangTableConstructorExpr tableConstructorExpr, AnalyzerData d
 
     @Override
     public void visit(BLangRecordLiteral recordLiteral, AnalyzerData data) {
+        BType referredType = Types.getReferredType(recordLiteral.getBType());
         List<RecordLiteralNode.RecordField> fields = recordLiteral.fields;
 
         for (RecordLiteralNode.RecordField field : fields) {
             if (field.isKeyValueField()) {
                 analyzeExpr(((BLangRecordKeyValueField) field).valueExpr, data);
+                reportIfDeprecatedUsage(((BLangRecordKeyValueField) field).key.fieldSymbol, recordLiteral,
+                        ((BLangRecordKeyValueField) field).pos);
             } else if (field.getKind() == NodeKind.SIMPLE_VARIABLE_REF) {
-                analyzeExpr((BLangRecordLiteral.BLangRecordVarNameField) field, data);
+                BLangRecordLiteral.BLangRecordVarNameField recField
+                        = (BLangRecordLiteral.BLangRecordVarNameField) field;
+                analyzeExpr(recField, data);
+
+                if (referredType.getKind() == TypeKind.RECORD) {
+                    BField matchingField = ((BRecordType) referredType).getFields().get(recField.symbol
```

Comment:
What I meant earlier was if `referredType` is a record assign `((BRecordType) referredType).getFields()` to a variable defined outside, or use null. Then you can just check for not null here?